File: go/core/execute.go
package core

import (
	"fmt"
	"strings"
)

// The executor — port of execute.ts / execute.rs, on the full-replacement state
// model. Handlers are invoked via closures (no reflection); panics are recovered
// (the assertion-style failure channel). State is a Value, replaced wholesale by
// each stimulus.

// StepOutcome is a step's outcome in the conformance trace.
type StepOutcome int

const (
	OutcomePass StepOutcome = iota
	OutcomeFail
	OutcomeSkipped
)

// String is the wire string ("pass"/"fail"/"skipped").
func (o StepOutcome) String() string {
	switch o {
	case OutcomePass:
		return "pass"
	case OutcomeFail:
		return "fail"
	default:
		return "skipped"
	}
}

// StepObservation is one executed step's outcome. ExampleIndex is 0-based;
// Ordinal is 1-based.
type StepObservation struct {
	ExampleIndex int
	Ordinal      int
	Outcome      StepOutcome
	Error        *StepFailure
}

// ExecutePorts are the ports the executor needs. CreateContext maps a step-file
// to its fresh initial state (nil → Null per file); Observer is optional
// per-step instrumentation; Reporter receives every planning diagnostic.
type ExecutePorts struct {
	Reporter      func(Diagnostic)
	CreateContext func(file string) any
	Observer      func(StepObservation)
}

// QueuedExample is one runnable example: its name and a callback that runs its steps.
type QueuedExample struct {
	Name string
	run  func() *StepFailure
}

// Run runs the example's steps; non-nil on the first failure.
func (q QueuedExample) Run() *StepFailure { return q.run() }

// CollectExamples reports every diagnostic in plan, then returns one
// QueuedExample per planned example, in document order (each run is lazy).
func CollectExamples(plan ExecutionPlan, ports ExecutePorts) []QueuedExample {
	if ports.Reporter != nil {
		for _, d := range plan.Diagnostics {
			ports.Reporter(d)
		}
	}
	queue := make([]QueuedExample, len(plan.Examples))
	for i := range plan.Examples {
		idx := i
		ex := plan.Examples[i]
		queue[i] = QueuedExample{
			Name: ex.Name,
			run:  func() *StepFailure { return runExample(plan, ex, idx, ports) },
		}
	}
	return queue
}

// ExecutePlan runs every example in plan, in order, stopping at the first failure.
func ExecutePlan(plan ExecutionPlan, ports ExecutePorts) *StepFailure {
	for _, q := range CollectExamples(plan, ports) {
		if f := q.Run(); f != nil {
			return f
		}
	}
	return nil
}

func runExample(plan ExecutionPlan, ex PlannedExample, exampleIndex int, ports ExecutePorts) *StepFailure {
	path := plan.VarDoc.Path
	source := plan.VarDoc.Source
	steps := ex.Steps

	stateByFile := map[string]any{}
	var lastReturn *Value
	var thrown *StepFailure

	for i := range steps {
		step := steps[i]
		file := step.StepDef.ExpressionSourceFile
		state, ok := stateByFile[file]
		if !ok {
			state = createContext(ports, file)
			stateByFile[file] = state
		}

		// A trailing data table / doc string is the last handler argument.
		callArgs := append([]Value{}, step.Args...)
		if step.DataTable != nil {
			callArgs = append(callArgs, tableRows(*step.DataTable))
		} else if step.DocString != nil {
			callArgs = append(callArgs, StrValue(step.DocString.Body))
		}

		var stepError *StepError
		returned, herr := invokeResolve(step.StepDef.Handler, state, callArgs)
		if herr != nil {
			e := handlerStepError(*herr)
			stepError = &e
		} else {
			lastReturn = asValuePointer(returned)
			switch {
			case step.StepDef.Kind == nil:
				e := returnShapeError("unknown step kind: null")
				stepError = &e
			case *step.StepDef.Kind == Stimulus:
				stateByFile[file] = returned
			case *step.StepDef.Kind == Sensor:
				// Header-bound rows are checked after the loop via RowChecks.
				if ex.RowChecks == nil {
					stepError = checkSensorReturn(source, step, asValuePointer(returned))
				}
			}
		}

		if stepError == nil {
			observe(ports, StepObservation{ExampleIndex: exampleIndex, Ordinal: i + 1, Outcome: OutcomePass})
		} else {
			failure := attachLocation(*stepError, step, path)
			observe(ports, StepObservation{ExampleIndex: exampleIndex, Ordinal: i + 1, Outcome: OutcomeFail, Error: &failure})
			thrown = &failure
			break
		}
	}

	// Header-bound row checks (deferred to after the loop).
	if thrown == nil && ex.RowChecks != nil && len(ex.RowChecks) > 0 {
		var bad []CellDiff
		for _, d := range compareRow(lastReturn, ex.RowChecks) {
			if !d.Ok {
				bad = append(bad, d)
			}
		}
		// Like a slotted sensor, a header-bound row step must answer the row it is
		// bound to: no return means nothing was compared.
		if lastReturn == nil || len(bad) > 0 {
			lastStep := steps[len(steps)-1]
			err := cellMismatchError(bad)
			if lastReturn == nil {
				err = returnShapeError("a header-bound row step must return a row object with one value per bound column, got nothing")
			}
			failure := attachLocation(err, lastStep, path)
			observe(ports, StepObservation{ExampleIndex: exampleIndex, Ordinal: len(steps), Outcome: OutcomeFail, Error: &failure})
			thrown = &failure
		}
	}

	// Error-fence inversion.
	if ex.ExpectedOutcome != nil && *ex.ExpectedOutcome == "fail" {
		if thrown == nil {
			if len(steps) > 0 {
				f := attachLocation(StepError{Kind: SEUnexpectedPass}, steps[len(steps)-1], path)
				return &f
			}
			f := bareFailure(StepError{Kind: SEUnexpectedPass})
			return &f
		}
		if ex.ExpectedErrorMessage != nil {
			if !strings.Contains(thrown.Error.Message(), *ex.ExpectedErrorMessage) {
				return thrown
			}
		}
		return nil
	}

	return thrown
}

func createContext(ports ExecutePorts, file string) any {
	if ports.CreateContext != nil {
		return ports.CreateContext(file)
	}
	return nil
}

// asValuePointer reads a handler result as a comparable Value. A stimulus's
// next state may be any Go value and is never read this way; a sensor's result
// is always a *Value (the facade boxes it), and nil means "no assertion".
func asValuePointer(result any) *Value {
	switch v := result.(type) {
	case nil:
		return nil
	case *Value:
		return v
	case Value:
		return &v
	}
	return nil
}

func observe(ports ExecutePorts, observation StepObservation) {
	if ports.Observer != nil {
		ports.Observer(observation)
	}
}

func tableRows(table Table) Value {
	row := func(cells []string) Value {
		vs := make([]Value, len(cells))
		for i, c := range cells {
			vs[i] = StrValue(c)
		}
		return ListOf(vs)
	}
	rows := []Value{row(table.Header.Cells)}
	for _, r := range table.Rows {
		rows = append(rows, row(r.Cells))
	}
	return ListOf(rows)
}

func attachLocation(error StepError, step PlannedStep, varPath string) StepFailure {
	a := anchor(error, step.MatchSpan)
	return StepFailure{
		Error: error,
		Location: &FailureLocation{
			Label: truncateLabel(step.Text),
			Path:  varPath,
			Line:  a.StartLine,
		},
	}
}

func truncateLabel(text string) string {
	if utf16Len(text) > 60 {
		runes := []rune(text)
		if len(runes) > 60 {
			runes = runes[:60]
		}
		return string(runes) + "…"
	}
	return text
}

func checkSensorReturn(source string, step PlannedStep, returned *Value) *StepError {
	extraCount := 0
	if step.DataTable != nil || step.DocString != nil {
		extraCount = 1
	}
	slotCount := len(step.Args) + extraCount
	if slotCount == 0 {
		// Nothing to compare against: returning nothing is the pass, a value is a mistake.
		if returned == nil {
			return nil
		}
		e := returnShapeError("this sensor has no parameters, data table or doc string — nothing to compare a return value against (throw to fail, return nothing to pass)")
		return &e
	}
	// With one or more slots the return is REQUIRED: returning nothing used to
	// skip the comparison silently, so a typo turned an assertion into a no-op.
	if returned == nil {
		e := returnShapeError(fmt.Sprintf("a sensor with %d slot(s) must return one value per slot, got nothing", slotCount))
		return &e
	}
	var slots []Value
	if slotCount == 1 {
		// The return IS the single slot's value, never read as a positional list.
		slots = []Value{*returned}
	} else {
		if returned.Kind != KindList {
			e := returnShapeError(fmt.Sprintf("a sensor with %d parameters must return a List of %d values, got %s", slotCount, slotCount, returned.TypeName()))
			return &e
		}
		if len(returned.List) != slotCount {
			e := returnShapeError(fmt.Sprintf("sensor return must have %d element(s), got %d", slotCount, len(returned.List)))
			return &e
		}
		slots = returned.List
	}

	argCount := len(step.Args)
	if argCount > 0 {
		sourceTexts := make([]string, len(step.ParamSpans))
		for i, s := range step.ParamSpans {
			sourceTexts[i] = utf16Slice(source, s.StartOffset, s.EndOffset)
		}
		var bad []CellDiff
		for _, d := range compareParamsWithFormats(slots[0:argCount], step.Args, step.ParamSpans, sourceTexts, step.Formats) {
			if !d.Ok {
				bad = append(bad, d)
			}
		}
		if len(bad) > 0 {
			e := cellMismatchError(bad)
			return &e
		}
	}

	if step.DataTable != nil {
		diffs, err := compareTable(&slots[argCount], *step.DataTable)
		if err != nil {
			return err
		}
		var bad []CellDiff
		for _, d := range diffs {
			if !d.Ok {
				bad = append(bad, d)
			}
		}
		if len(bad) > 0 {
			e := cellMismatchError(bad)
			return &e
		}
	} else if step.DocString != nil {
		diff, err := compareDocString(&slots[argCount], step.DocString.Body, step.DocString.BodySpan)
		if err != nil {
			return err
		}
		if diff != nil {
			e := docStringMismatchError(*diff)
			return &e
		}
	}
	return nil
}

// invokeResolve invokes the handler, normalising a returned error — and
// recovering a panic (the assertion-style failure channel) — into a
// HandlerError. Returns (returned, nil) on success where returned is nil for
// "no value", or (nil, err) on failure.
func invokeResolve(handler Handler, state any, args []Value) (returned any, herr *HandlerError) {
	defer func() {
		if r := recover(); r != nil {
			returned = nil
			herr = handlerErrorFromPanic(r)
		}
	}()
	v, err := handler.call(state, args)
	if err != nil {
		return nil, NewHandlerError(err.Error())
	}
	return v, nil
}

func handlerErrorFromPanic(r any) *HandlerError {
	switch v := r.(type) {
	case *HandlerError:
		return v
	case error:
		return NewHandlerError(v.Error())
	case string:
		return NewHandlerError(v)
	default:
		return NewHandlerError(fmt.Sprintf("%v", r))
	}
}

File: rust/core/src/execute.rs
//! The executor — port of `execute.ts` / `Execute.java`, on the full-replacement
//! state model. Handlers are invoked via boxed closures (no reflection); panics
//! are caught (the `AssertionError`/`Throwable` parity channel); `Future`
//! returns are driven by a small std `block_on`. State is a [`Value`], replaced
//! wholesale by each stimulus.

use crate::cell_diff::{CellDiff, compare_row, compare_table};
use crate::diagnostics::Diagnostic;
use crate::doc_string_diff::compare_doc_string;
use crate::error::{FailureLocation, HandlerError, StepError, StepFailure};
use crate::failure_anchor;
use crate::handler::{Handler, StepOutput, StepReturn};
use crate::offsets::{utf16_len, utf16_slice};
use crate::param_diff::compare_params_with_formats;
use crate::plan::{ExecutionPlan, PlannedExample, PlannedStep};
use crate::step_kind::StepKind;
use crate::value::Value;
use std::any::Any;
use std::cell::Cell;
use std::collections::HashMap;
use std::future::Future;
use std::panic::AssertUnwindSafe;
use std::pin::Pin;
use std::rc::Rc;
use std::sync::Once;
use std::task::{Context, Poll, Wake, Waker};

/// A step's outcome in the conformance trace.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum StepOutcome {
    Pass,
    Fail,
    Skipped,
}

impl StepOutcome {
    /// The wire string (`"pass"`/`"fail"`/`"skipped"`).
    pub fn as_str(self) -> &'static str {
        match self {
            StepOutcome::Pass => "pass",
            StepOutcome::Fail => "fail",
            StepOutcome::Skipped => "skipped",
        }
    }
}

/// One executed step's outcome. `example_index` is 0-based; `ordinal` is 1-based.
#[derive(Clone, Debug, PartialEq)]
pub struct StepObservation {
    pub example_index: usize,
    pub ordinal: usize,
    pub outcome: StepOutcome,
    pub error: Option<StepFailure>,
}

/// The ports [`collect_examples`]/[`execute_plan`] need. `create_context` maps a
/// step-file to its fresh initial state (`None` → a unit state per file);
/// `observer` is optional per-step instrumentation. The lifetime lets the port
/// closures borrow caller locals (e.g. a conformance observer's accumulator).
pub struct ExecutePorts<'a> {
    pub reporter: Reporter<'a>,
    pub create_context: Option<ContextFactory<'a>>,
    pub observer: Option<Observer<'a>>,
}

/// Receives every diagnostic collected during planning.
pub type Reporter<'a> = Box<dyn Fn(&Diagnostic) + 'a>;
/// Maps a step-file to its fresh initial state.
pub type ContextFactory<'a> = Box<dyn Fn(&str) -> Rc<dyn Any> + 'a>;
/// Per-step instrumentation (conformance trace mode).
pub type Observer<'a> = Box<dyn Fn(StepObservation) + 'a>;

impl<'a> ExecutePorts<'a> {
    /// Ports with just a reporter (no context factory, no observer).
    pub fn new(reporter: Box<dyn Fn(&Diagnostic) + 'a>) -> ExecutePorts<'a> {
        ExecutePorts {
            reporter,
            create_context: None,
            observer: None,
        }
    }
}

impl ExecutePorts<'static> {
    /// Ports that discard diagnostics and observe nothing.
    pub fn silent() -> ExecutePorts<'static> {
        ExecutePorts::new(Box::new(|_| {}))
    }
}

/// One runnable example: its name and a callback that runs its steps.
pub struct QueuedExample<'a> {
    pub name: String,
    run: Box<dyn Fn() -> Result<(), StepFailure> + 'a>,
}

impl QueuedExample<'_> {
    /// Runs the example's steps; `Err` on the first failure.
    pub fn run(&self) -> Result<(), StepFailure> {
        (self.run)()
    }
}

/// Reports every diagnostic in `plan`, then returns one [`QueuedExample`] per
/// planned example, in document order (each `run` is lazy). Port of `collectExamples`.
pub fn collect_examples<'a>(
    plan: &'a ExecutionPlan,
    ports: &'a ExecutePorts<'a>,
) -> Vec<QueuedExample<'a>> {
    for d in &plan.diagnostics {
        (ports.reporter)(d);
    }
    plan.examples
        .iter()
        .enumerate()
        .map(|(i, ex)| QueuedExample {
            name: ex.name.clone(),
            run: Box::new(move || run_example(plan, ex, i, ports)),
        })
        .collect()
}

/// Runs every example in `plan`, in order, stopping at the first failure. Port
/// of `executePlan`.
pub fn execute_plan<'a>(
    plan: &'a ExecutionPlan,
    ports: &'a ExecutePorts<'a>,
) -> Result<(), StepFailure> {
    for q in collect_examples(plan, ports) {
        q.run()?;
    }
    Ok(())
}

// -----------------------------------------------------------------------------
// One example
// -----------------------------------------------------------------------------

fn run_example(
    plan: &ExecutionPlan,
    ex: &PlannedExample,
    example_index: usize,
    ports: &ExecutePorts,
) -> Result<(), StepFailure> {
    let path = &plan.var_doc.path;
    let source = &plan.var_doc.source;
    let steps = &ex.steps;

    let mut state_by_file: HashMap<String, Rc<dyn Any>> = HashMap::new();
    let mut last_return: Option<Value> = None;
    let mut thrown: Option<StepFailure> = None;

    for (i, step) in steps.iter().enumerate() {
        let file = &step.step_def.expression_source_file;
        let state = match state_by_file.get(file) {
            Some(s) => s.clone(),
            None => {
                let created = create_context(ports, file);
                state_by_file.insert(file.clone(), created.clone());
                created
            }
        };

        // A trailing data table / doc string is the last handler argument.
        let mut call_args = step.args.clone();
        if let Some(table) = &step.data_table {
            call_args.push(table_rows(table));
        } else if let Some(fence) = &step.doc_string {
            call_args.push(Value::from(fence.body.as_str()));
        }

        let step_error: Option<StepError> =
            match invoke_resolve(&step.step_def.handler, state, call_args) {
                Err(he) => Some(StepError::Handler(he)),
                Ok(output) => {
                    last_return = output.compared().cloned();
                    match step.step_def.kind {
                        Some(StepKind::Stimulus) => {
                            // A stimulus's output IS the next state (full
                            // replacement). The facade yields `State`; the core's
                            // Value-state conveniences yield `Compared`, which for
                            // a stimulus means the same thing. Returning nothing
                            // (`Compared(None)`) leaves state unchanged — it is not
                            // the same as returning `Value::Null`.
                            let next: Option<Rc<dyn Any>> = match output {
                                StepOutput::State(next) => Some(next),
                                StepOutput::Compared(v) => v.map(|v| Rc::new(v) as Rc<dyn Any>),
                            };
                            if let Some(next) = next {
                                state_by_file.insert(file.clone(), next);
                            }
                            None
                        }
                        Some(StepKind::Sensor) => {
                            // Header-bound rows are checked after the loop via row_checks.
                            if ex.row_checks.is_none() {
                                check_sensor_return(source, step, output.compared().cloned()).err()
                            } else {
                                None
                            }
                        }
                        None => Some(StepError::ReturnShape("unknown step kind: null".to_string())),
                    }
                }
            };

        match step_error {
            None => observe(
                ports,
                StepObservation {
                    example_index,
                    ordinal: i + 1,
                    outcome: StepOutcome::Pass,
                    error: None,
                },
            ),
            Some(err) => {
                let failure = attach_location(err, step, path);
                observe(
                    ports,
                    StepObservation {
                        example_index,
                        ordinal: i + 1,
                        outcome: StepOutcome::Fail,
                        error: Some(failure.clone()),
                    },
                );
                thrown = Some(failure);
                break;
            }
        }
    }

    // Header-bound row checks (deferred to after the loop).
    if thrown.is_none() {
        if let Some(checks) = &ex.row_checks {
            if !checks.is_empty() {
                let bad: Vec<CellDiff> = compare_row(last_return.as_ref(), checks)
                    .into_iter()
                    .filter(|d| !d.ok)
                    .collect();
                if !bad.is_empty() {
                    let last_step = steps.last().unwrap();
                    let failure = attach_location(StepError::CellMismatch(bad), last_step, path);
                    observe(
                        ports,
                        StepObservation {
                            example_index,
                            ordinal: steps.len(),
                            outcome: StepOutcome::Fail,
                            error: Some(failure.clone()),
                        },
                    );
                    thrown = Some(failure);
                }
            }
        }
    }

    // Error-fence inversion.
    if ex.expected_outcome.as_deref() == Some("fail") {
        match thrown {
            None => {
                return Err(match steps.last() {
                    Some(last) => attach_location(StepError::UnexpectedPass, last, path),
                    None => StepFailure::bare(StepError::UnexpectedPass),
                });
            }
            Some(failure) => {
                if let Some(expected_msg) = &ex.expected_error_message {
                    if !failure.error.message().contains(expected_msg) {
                        return Err(failure);
                    }
                }
                return Ok(());
            }
        }
    }

    match thrown {
        Some(failure) => Err(failure),
        None => Ok(()),
    }
}

fn create_context(ports: &ExecutePorts, file: &str) -> Rc<dyn Any> {
    match &ports.create_context {
        Some(cc) => cc(file),
        None => Rc::new(()) as Rc<dyn Any>,
    }
}

fn observe(ports: &ExecutePorts, observation: StepObservation) {
    if let Some(observer) = &ports.observer {
        observer(observation);
    }
}

fn table_rows(table: &crate::ast::Table) -> Value {
    let row =
        |cells: &[String]| Value::List(cells.iter().map(|c| Value::from(c.as_str())).collect());
    let mut rows = vec![row(&table.header.cells)];
    for r in &table.rows {
        rows.push(row(&r.cells));
    }
    Value::List(rows)
}

fn attach_location(error: StepError, step: &PlannedStep, var_path: &str) -> StepFailure {
    let anchor = failure_anchor::anchor(&error, step.match_span);
    let label = truncate_label(&step.text);
    StepFailure {
        error,
        location: Some(FailureLocation {
            label,
            path: var_path.to_string(),
            line: anchor.start_line,
        }),
    }
}

fn truncate_label(text: &str) -> String {
    if utf16_len(text) > 60 {
        let truncated: String = text.chars().take(60).collect();
        format!("{truncated}…")
    } else {
        text.to_string()
    }
}

// -----------------------------------------------------------------------------
// Sensor return comparison
// -----------------------------------------------------------------------------

fn check_sensor_return(
    source: &str,
    step: &PlannedStep,
    returned: Option<Value>,
) -> Result<(), StepError> {
    let Some(returned) = returned else {
        return Ok(());
    };
    let extra_count = usize::from(step.data_table.is_some() || step.doc_string.is_some());
    let slot_count = step.args.len() + extra_count;
    if slot_count == 0 {
        return Err(StepError::ReturnShape(
            "this sensor has no parameters, data table or doc string — nothing to compare a return value against \
             (throw to fail, return nothing to pass)"
                .to_string(),
        ));
    }
    let slots: Vec<Value> = if slot_count == 1 {
        // The return IS the single slot's value, never read as a positional list.
        vec![returned]
    } else {
        match returned {
            Value::List(list) => {
                if list.len() != slot_count {
                    return Err(StepError::ReturnShape(format!(
                        "sensor return must have {} element(s), got {}",
                        slot_count,
                        list.len()
                    )));
                }
                list
            }
            other => {
                return Err(StepError::ReturnShape(format!(
                    "a sensor with {} parameters must return a List of {} values, got {}",
                    slot_count,
                    slot_count,
                    other.type_name()
                )));
            }
        }
    };

    let arg_count = step.args.len();
    if arg_count > 0 {
        let source_texts: Vec<String> = step
            .param_spans
            .iter()
            .map(|s| utf16_slice(source, s.start_offset, s.end_offset).to_string())
            .collect();
        let bad: Vec<CellDiff> = compare_params_with_formats(
            &slots[0..arg_count],
            &step.args,
            &step.param_spans,
            &source_texts,
            Some(&step.formats),
        )
        .into_iter()
        .filter(|d| !d.ok)
        .collect();
        if !bad.is_empty() {
            return Err(StepError::CellMismatch(bad));
        }
    }

    if let Some(table) = &step.data_table {
        let bad: Vec<CellDiff> = compare_table(Some(&slots[arg_count]), table)?
            .into_iter()
            .filter(|d| !d.ok)
            .collect();
        if !bad.is_empty() {
            return Err(StepError::CellMismatch(bad));
        }
    } else if let Some(fence) = &step.doc_string {
        if let Some(diff) =
            compare_doc_string(Some(&slots[arg_count]), &fence.body, fence.body_span)?
        {
            return Err(StepError::DocStringMismatch(diff));
        }
    }
    Ok(())
}

// -----------------------------------------------------------------------------
// Handler invocation (panic-catching + async resolution)
// -----------------------------------------------------------------------------

thread_local! {
    static SUPPRESS_PANIC: Cell<bool> = const { Cell::new(false) };
}

static HOOK: Once = Once::new();

/// Installs a panic hook (once) that suppresses the default stderr print for
/// panics the executor deliberately catches (a handler's assertion-style
/// failure), while leaving genuine test panics untouched on other threads.
///
/// DECLARED EXCEPTION to the "no globals in the core" rule (see `lib.rs`):
/// `catch_unwind` is the executor's assertion channel — the AssertionError
/// parity with Java — and the process-wide hook is the only way Rust offers to
/// keep a *caught* panic from spewing to stderr. It is `Once`-guarded, chains
/// the previous hook, and gates on a thread-local so it is inert outside
/// [`invoke_resolve`]; observable behaviour is otherwise unchanged.
fn install_hook() {
    HOOK.call_once(|| {
        let previous = std::panic::take_hook();
        std::panic::set_hook(Box::new(move |info| {
            if SUPPRESS_PANIC.with(Cell::get) {
                return;
            }
            previous(info);
        }));
    });
}

/// Invokes the handler and resolves any `Future`, catching a panic (the
/// assertion-style failure channel) into a [`HandlerError`].
fn invoke_resolve(
    handler: &Handler,
    state: Rc<dyn Any>,
    args: Vec<Value>,
) -> Result<StepOutput, HandlerError> {
    install_hook();
    let caught = SUPPRESS_PANIC.with(|s| {
        s.set(true);
        let r = std::panic::catch_unwind(AssertUnwindSafe(|| match handler.call(state, args) {
            StepReturn::Ready(r) => r,
            StepReturn::Pending(fut) => block_on(fut),
        }));
        s.set(false);
        r
    });
    match caught {
        Ok(r) => r,
        Err(payload) => Err(HandlerError::from_panic(payload)),
    }
}

/// A minimal `block_on`: polls the future, parking the thread until its waker
/// unparks it. No dependencies, no unsafe.
fn block_on<T>(mut fut: Pin<Box<dyn Future<Output = T>>>) -> T {
    struct ThreadWaker(std::thread::Thread);
    impl Wake for ThreadWaker {
        fn wake(self: std::sync::Arc<Self>) {
            self.0.unpark();
        }
        fn wake_by_ref(self: &std::sync::Arc<Self>) {
            self.0.unpark();
        }
    }
    let waker = Waker::from(std::sync::Arc::new(ThreadWaker(std::thread::current())));
    let mut cx = Context::from_waker(&waker);
    loop {
        match fut.as_mut().poll(&mut cx) {
            Poll::Ready(v) => return v,
            Poll::Pending => std::thread::park(),
        }
    }
}
